src/Microsoft.DotNet.Cli.Utils/LineNotificationTextWriter.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    /// <summary>
+    /// A TextWriter that can raises an event for each line that is written to it.
+    /// </summary>
+    public class LineNotificationTextWriter : TextWriter
+    {
+        private Encoding _encoding;
+        private StringBuilder _currentString;
+        private Action<string> _lineHandler;
+
+        public LineNotificationTextWriter(IFormatProvider formatProvider, Encoding encoding)
+            : base(formatProvider)
+        {
+            _encoding = encoding;
+
+            // start with an average line length so the builder doesn't need to immediately grow
+            _currentString = new StringBuilder(128);
+        }
+
+        public LineNotificationTextWriter OnWriteLine(Action<string> lineHandler)
+        {
+            if (lineHandler == null)
+            {
+                throw new ArgumentNullException(nameof(lineHandler));
+            }
+            if (_lineHandler != null)
+            {
+                throw new InvalidOperationException("OnWriteLine has already been set.");
+            }
+
+            _lineHandler = lineHandler;
+
+            return this;
+        }
+
+        public override Encoding Encoding
+        {
+            get { return _encoding; }
+        }
+
+        // Write(char) gets called for all overloads of Write
+        public override void Write(char value)
+        {
+            lock (_currentString)
```

Comment:
It is for the case where a BuiltInCommand uses multiple threads that write to Console.Out or Console.Error. It probably doesn't happen currently, but if this starts getting reused in other places, or compile-csc starts doing more things, it felt better to be safe since this object is going to be held in the static Console.Out property.